🚀 4단계 - 자동차 경주(우승자)

sjjeong opened the pull request:
안녕하세요.
1년 반 만에 과제를 다시 하러 왔습니다 ㅎㅎ 잘 부탁드립니다 🙏 

자동차 경주(우승자) 과제 진행했습니다.

Engine 관련 리뷰 반영했습니다: https://github.com/next-step/kotlin-racingcar/pull/1082#discussion_r1229364856
joinToString 관련 리뷰 반영했습니다: https://github.com/next-step/kotlin-racingcar/pull/1082#discussion_r1229361683

sjjeong committed:
feat: 자동차 이름을 입력 받도록 변경

sjjeong committed:
feat: 자동차 이름을 출력하게 변경

sjjeong committed:
feat: 승자들이 출력되게 변경

sjjeong committed:
refactor: 자동차 움직이는 동작을 CarEngine으로 변경

sjjeong committed:
refactor: showProcess의 출력 로직 개선

StringBuilder를 사용하던 코드에서 joinToString으로 변경

sjjeong committed:
feat: 자동차 이름은 5자를 초과할 수 없다

ddaaac commented on src/main/kotlin/study/step3/CarEngine.kt:
```diff
@@ -0,0 +1,18 @@
+    companion object {
+        const val MIN_RPM_TO_MOVE = 4
+    }
+}
```
intelij에서는 파일 마지막에 자동으로 [개행을 추가](https://velog.io/@d-h-k/intellij-%ED%8C%8C%EC%9D%BC%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EC%B6%94%EA%B0%80%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95)할 수 있습니다!

sjjeong replied:
8e5274e771dac6824b360b986d9894bffa988eab

ddaaac commented on src/main/kotlin/study/step3/Mileage.kt:
```diff
@@ -0,0 +1,8 @@
+package study.step3
+
+@JvmInline
+value class Mileage(val value: Int) {
```
value class를 사용해주셨군요 💯

ddaaac commented on src/main/kotlin/study/step3/Score.kt:
```diff
@@ -1,4 +1,6 @@
-@JvmInline
-value class Score(val score: Int)
+data class Score(
+    val carName: String,
```
Score가 name을 갖고 있는 구조가 어색한데, 이름이 필요하다면 Car에서 가져올 수 있지 않을까요?

sjjeong replied:
Score 라는 이름이 가지는 의미를 생각해봤을때 지금은 필요 없을 것 같아서 제거하고 Scoreboard.Item으로 변경해봤습니다
5d5ee01555d0bc522683c07cbbc3fed974aa45df

ddaaac commented on src/main/kotlin/study/step3/Scoreboard.kt:
```diff
@@ -8,18 +8,25 @@ class Scoreboard {
+    private fun showWinners() {
+        val finalScores = scoresList.lastOrNull() ?: return
+        val maxScore: Mileage = finalScores.maxBy { it.score.value }.score
+        val winnerNames: String = finalScores.filter { it.score == maxScore }.joinToString { it.carName }
```
우승자를 구하는 로직을 테스트해보고 싶은 생각이 드셨을 거 같아요!
이러한 **로직**을 테스트하려면 View가 아닌 도메인에 있어야하지 않을까요?
출력 관련된 로직들은 모두 View로 보내고 ScoreBoard는 도메인으로 남기면 좋을 것 같습니다~

sjjeong replied:
로직 분리를 해봤습니다!
1ffc581945af2db0600ca32c12ea3bef6989fe09

ddaaac reviewed (changes requested):
안녕하세요 석준님 👋 

구현을 잘 해주셔서 크게 남길 코멘트는 없네요.

코멘트 확인 부탁드려요!

sjjeong committed:
refactor: 마지막 줄에 개행 추가

sjjeong committed:
refactor: entity 패키지 추가

entity 패키지에 관련 클래스들을 이동

sjjeong committed:
refactor: CarEngine 이름 변경

CarEngine 보다 Engine이 더 잘 어울릴 것 같음

sjjeong committed:
refactor: Scoreboard 함수 정리

Scoreboard가 Output의 역할을 하고 있어서 함수들을 정리함

sjjeong committed:
refactor: Score를 제거

Score 자체는 점수를 나타내는 것인데 name이 들어가다 보니 클래스가 어색해짐

Scoreboard에 들어가는 Item라는 의미로 이름을 변경하고 Scoreboard는 Item을 가지는 List를 가지는 Collection의 의미로 변경

ddaaac commented on src/main/kotlin/study/step3/entity/Engine.kt:
```diff
@@ -0,0 +1,18 @@
+    private val minPerToMove: Int = MIN_RPM_TO_MOVE
+) {
+
+    fun accelerate(rpm: Int): Mileage {
```
추상화를 하는 정도나, 코드를 현실과 일치시키는 정도 모두 쉽지 않은 것 같아요.
개인적으로는 어느 정도의 직관성을 가져가면 좋을 것 같습니다.
`Engine#accelerate`를 봤을 때는 어떤 역할을 할 지 바로 가늠이 되지 않는 것 같아요. `goOrStop`등의 이름은 어떻게 생각하실까요?

ddaaac commented on src/main/kotlin/study/step3/entity/Car.kt:
```diff
@@ -0,0 +1,27 @@
+
+data class Car(
+    val name: String,
+    val engine: Engine,
+) {
+    var mileage: Mileage = Mileage(0)
+        private set
+
+    init {
+        check(name.length <= MAX_CAR_NAME_LENGTH) {
+            "자동차 이름은 5자를 초과할 수 없습니다."
+        }
+    }
+
+    fun move() {
+        val rpm: Int = Random.nextInt(MAX_RPM + 1)
+        mileage += engine.accelerate(rpm = rpm)
```
그러고보니 든 생각인데 지금 구조에서 engine과 Car 객체를 나눌 필요가 있을까요?
두 객체가 하는 일이 크게 다른 것 같지 않아서 합쳐도 될 것 같아보이네요.

ddaaac commented on src/main/kotlin/study/step3/entity/Scoreboard.kt:
```diff
@@ -0,0 +1,27 @@
+package study.step3.entity
+
+
+class Scoreboard : MutableList<List<Scoreboard.Item>> by mutableListOf() {
```
List 대신 MutableList를 구현해주신 이유가 있을까요?

ddaaac commented on src/main/kotlin/study/step3/entity/Scoreboard.kt:
```diff
@@ -0,0 +1,27 @@
+        get() {
+            return map { scores ->
+                scores.joinToString("\n") { item ->
+                    "${item.title} ${"-".repeat(item.score)}"
```
출력에 관한 로직은 View에서 다루는 건 어떨까요?

ddaaac commented on src/main/kotlin/study/step3/entity/Scoreboard.kt:
```diff
@@ -0,0 +1,27 @@
+            val finalScores = lastOrNull() ?: return emptyList()
+            val maxScore: Int = finalScores.maxBy { it.score }.score
+            return finalScores.filter { it.score == maxScore }
+                .map { it.title }
```
여기도 마찬가지로 어떤 값을 출력할지는 view에서 결정해도 좋아보여요.

ddaaac commented on src/test/kotlin/study/step3/ScoreboardTest.kt:
```diff
@@ -0,0 +1,45 @@
+import study.step3.entity.Scoreboard
+
+class ScoreboardTest {
+    private lateinit var scoreboard: Scoreboard
+
+    @BeforeEach
+    fun setUp() {
+        scoreboard = Scoreboard()
+    }
+
+    @Test
+    fun `최고점이 1명이면 승자는 1명이다`() {
+        scoreboard.add(
```
각 테스트에서 val scoreBoard = ScoreBoard() 처럼 초기화해도 되지 않을까요?

ddaaac reviewed (approved):
안녕하세요 석준님 👋 

리뷰 반영 잘 해주셨습니다.

다음 단계에서 더 얘기 나눠볼게요!
